src/main/lock.ts

```diff
@@ -0,0 +1,44 @@
+// ==LICENSE-BEGIN==
+// Copyright 2017 European Digital Reading Lab. All rights reserved.
+// Licensed to the Readium Foundation under one or more contributor license agreements.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file exposed on Github (readium) in the project repository.
+// ==LICENSE-END==
+
+import * as debug_ from "debug";
+import { app } from "electron";
+import { container } from "readium-desktop/main/di";
+import { WinRegistry } from "readium-desktop/main/services/win-registry";
+
+import { cli } from "./cli/process";
+
+// Logger
+const debug = debug_("readium-desktop:main:lock");
+
+export function lockInstance() {
+    const gotTheLock = app.requestSingleInstanceLock();
+
+    if (gotTheLock) {
+        app.on("second-instance", (_e, commandLine, _workingDir) => {
```

Comment:
The `second-instance` event handling should be side-by-side with the MacOS `open-file` event handling (within `will-finish-launching`), because they are logically-related.

You can see in this comment how I illustrated the different code paths, using `handleFilePath()` and `handleArgv()` (which itself can ultimately invoke `handleFilePath()`, depending on the CLI parameters):
https://github.com/readium/readium-desktop/issues/547#issue-479589133

Reply:
Note that on MacOS, `open-file` is mutually-exclusive with `second-instance`, in the sense that the former is called when:
* double-clicking / activating a file in the Finder with an extension or content-type that is declaratively handled / registered by the app
* drag-and-drop of a file onto the application's icon
* AppleScript launch
...whereas the latter is called only when the application is launched via the command line (not with the `open` shell utility, but directly from the executable)

I tried to illustrate this in my example:
https://github.com/readium/readium-desktop/issues/547#issue-479589133

Reply:
Just to re-iterate, the `will-finish-launching` + `open-file` MacOS event handler must reside side-by-side (at the code level) with the `second-instance` event handler, because they are logically-related in the Electron app design pattern, which boils down to:

The latter event handler (`second-instance`) is invoked on Windows and Linux for app launches (i.e. CLI, or other types such as "open with" or "double-click on registered/supported file extension from explorer") that fail to acquire a "single instance lock", and only called on MacOS for direct CLI executions (which exclude the `open APP_NAME` or AppleScript app launches). The former (`open-file`) is therefore invoked for all other MacOS app launches that fail to acquire the "single instance lock" (which include the file drag-drop onto the app icon, etc.).

The solution is quite simple, move the event handler from `src/main.ts` to `src/main/lock.ts`:

https://github.com/readium/readium-desktop/blob/dc23718ed6e9e7ccbe69f7790e63b7b21669f63c/src/main.ts#L88-L100